### ISSUE TITLE
Integrate auto-generated CLI/API documentation for `dbt-core`

### ISSUE THREAD

**Issue by jtcohen6** [fishtown-analytics/docs.getdbt.com#2674]
### Contributions

- [X] I have read the contribution docs, and understand what's expected of me.

### What page(s) or areas on docs.getdbt.com are affected?

https://docs.getdbt.com/reference/dbt-commands
https://docs.getdbt.com/reference/global-cli-flags / https://docs.getdbt.com/reference/global-configs
?? https://docs.getdbt.com/docs/deploy/deployments

### What changes are you suggesting?

The Core team is in progress on a new CLI + "top-level" API for `dbt-core`, which will enable both:
1. Better automated documentation of commands + parameters (flags / arguments)
2. Programmatic invocations of `dbt-core`, at parity with CLI invocations — allowing us to finally update our docs on `dbt-core`'s "Python API" (https://github.com/dbt-labs/docs.getdbt.com/issues/1973). While this is a **huge and impactful** change, it's out of scope for the current issue, since it's just a question of updating content, not requiring engineering effort.

## Auto-generated documentation

The big idea is, these docs are generated using docstrings defined _within_ the `dbt-core` codebase. They're still written by human beings, but it's the same human beings who are designing, developing, & changing `dbt-core` — rather than a separate set of human beings, in a separate repository, after the fact.

Here's a sample of how it looks right now, on a feature branch of `dbt-core`:
<img width="931" alt="Screenshot 2023-01-08 at 20 06 39" src="https://user-images.githubusercontent.com/13897643/211214203-dedeeb98-6128-454a-b678-e7804ca0ccc8.png">

These docs are generated using [Sphinx](https://www.sphinx-doc.org/en/master/index.html), the standard for Python projects, and it outputs ready-to-read HTML.

Things we'd need to figure out:
- [ ] **Integration with Docusaurus.** The output of Sphinx is HTML, not MDX (Markdown + React).
- [ ] **Coordinating styles.** Sphinx can accept [custom CSS files](https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-html_css_files), but the ideal approach would allow us to specify styles only once in docs.getdbt.com.
- [ ] **Automation & versioning.** I'd anticipate us generating the docs once per minor version, and then directing users to the appropriate one based on their selection in the version dropdown. Whenever we release a new minor or patch version of `dbt-core`, we'd need some automation (PR?) to update the generated docs for that minor version.
- [ ] **Content: overlap & discrepancies.** As a general rule, we should prefer to have as much relevant documentation as possible located right within the `dbt-core` codebase. There might be lengthier "narrative" docs or working examples that make more sense to be easily editable from within this repository, though—we could consider linking to those from within the `dbt-core` API docs, so long as we're good about our link stability & redirects. (This one can also be out of scope for the current issue, since it's about content, not engineering effort to integrate.)

### Additional information

https://github.com/dbt-labs/dbt-core/issues/6548

> It would be significantly better to integrate these docs directly with docs.getdbt.com, especially for pages that are now more-or-less duplicative.

https://www.notion.so/dbtlabs/dbt-Core-v1-5-docs-a2ae11be77bf47629f85faf7a3f27991

**Comment by jtcohen6** [comment]
Auto-gen API docs for `dbt-server` as well: https://github.com/dbt-labs/dbt-server/issues/153

Let's see if we can coordinate the two!

**Comment by runleonarun** [comment]
Thanks for writing this up @jtcohen6! Some comments/questions:
* I agree that most content should live in the core codebase. It's industry standard for APIs and makes sense being there so it can be added/edited when a feature is added/updated. 
* Can the docs team edit the API content/descriptions in `dbt-core`?
* Did we already generate a spec from our API? Are we using this spec to describe the API and generate docs?

**Comment by jtcohen6** [comment]
> Can the docs team edit the API content/descriptions in `dbt-core`?

Yes! This would require changes to the `dbt-core` repository (going through a git workflow), and the changes would be tied to / released in a specific version of `dbt-core`. This could be a good heuristic for deciding what information should live where:
- Specific syntax / usage instructions in the `dbt-core` codebase
- Evergreen or narrative content (higher-level than "write exactly X code to do Y thing") could continue to live in this (docs) codebase
- Depending on how we host/deploy these docs, we could link between them

> Did we already generate a spec from our API? Are we using this spec to describe the API and generate docs?

For `dbt-server`, which is built FastAPI, we get an OpenAPI spec "for free"! It can be popped into any Swagger-like UI and visualized, ez.

For `dbt-core`: We're not using a standard spec. My sense is that there isn't a good standard for Python projects / CLI tools, in the same way that OpenAPI makes this very easy for RESTful web APIs. Instead, we're using some code [here](https://github.com/dbt-labs/dbt-core/tree/main/core/dbt/docs/source) that leverages `sphinx` (standard Python documentation generator) + `click` (new CLI framework), to crawl through the `dbt-core` codebase and output HTML. We rebuild the docs on changes to relevant source files.

The good news: Sphinx is pretty flexible in terms of its output format. There's a default HTML rendering, but it can also (with extensions) output other formats, including plain old Markdown. I took this for a quick spin, and I think the results are quite promising: https://github.com/dbt-labs/docs.getdbt.com/pull/2693

<img width="1434" alt="Screenshot 2023-01-11 at 11 37 46" src="https://user-images.githubusercontent.com/13897643/211785115-748f7c85-23be-4021-88ca-fe552e54f945.png">



(cc @stu-k - would love to hear your thoughts here, too!)

**Comment by jtcohen6** [comment]
I scoped down my ambitions for this work, and I'm proposing some hand-written content for v1.5 / for now in:
- https://github.com/dbt-labs/docs.getdbt.com/pull/3118

**Comment by mirnawong1** [comment]
hey team, is this something still on the radar or should we close this?

**Comment by mirnawong1** [comment]
hey team, looks like the linked core issue is listed as not planned so am going to close this out. however if we plan to pick this up again, we can re-open!
